test: integrate pinny on stacks-node

fdefelici opened the pull request:
### Description

Integrate [pinny](https://github.com/BitcoinL2-Labs/pinny-rs.git) crate on `stacks-node` package.

Added a demo configuration about tag labels and tagged tests.

Here an example of test filtering  and execution using `cargo test`:

- run "slow" tests
  ```bash
  cargo test -p stacks-node :slow:
  ...
  running 2 tests
  
  test tests::signer::v0::block_proposal_rejection::t::slow::signer::bitcoind::t ... ignored
  test tests::integrations::contract_stx_transfer::t::slow::t ... ok
  
  test result: ok. 1 passed; 0 failed; 1 ignored; 0 measured; 251 filtered out; finished in 26.05s
  ```
- run "fast" tests
  ```bash
  cargo test -p stacks-node :fast:
  
  running 1 tests
  
  test tests::signer::v0::block_proposal_rejection::t::fast::signer::bitcoind::t ... ok
  
  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 252 filtered out; finished in 0.14s
  ```

### Applicable issues

- fixes #6033

### Additional info (benefits, drawbacks, caveats)

#### Drawback
In specific cases, because of test function mangling is not possible to use `super::<symbol>` within a test function. Full description of this drawbacks can be found here: https://github.com/BitcoinL2-Labs/pinny-rs/blob/master/README.md.
(easy fixable as proposed in the current change, importing the symbol properly outside the test function)

#### Tag Naming
~~One thing to notice is the tag "signer". By the fact it is not unique due to the module path where "signer" is a real module name,   we have 2 options:~~
- ~~filtering test using anti-clash mechanism using expression (like t::*:signer). Note: not possible with `cargo test`, but possible with nextest~~
- ~~Make the tag "unique" (e.g. "_signer", "tsigner").~~

_**EDIT1**: Maybe not a an issue because if filter by "signer" probably one wants to execute all the tests related to signers anyway._

_**EDIT2**: here follow a tag proposal, after some reasoning_
#### Tags Proposal
List of tags (non-exhaustive) that we could use, with related meaning and potential benefits:

| Tag             | Description                                  | Benefits                      |
|-----------------|----------------------------------------------|--------------------------------|
| `fast`(1)       | tests running withing a minute               | easy filtering tests opposed to `slow`. With nextest we could write `'!(test(name =~ "slow"))'` expression |
| `slow`          | tests running over a minute                  | avoid the use of `#[ignore]` as described in contributing page |
| `bitcoind`      | tests requiring bitcoin daemon               | avoid the use of `#[ignore]` and the code check for `BITCOIND_TEST` env variable within the test |
| `flaky`         | tests that are consistently flaky            | highlighting flaky tests and make them filter friendly |
| `sequential`(2) | tests requiring sequential execution         | avoid the use of `#[ignore]` as described in contributing page |
| `parallel`(1)   | tests supporting parallel execution          | easy filtering tests opposed to `sequential`. With nextest we could write `'!(test(name =~ "sequential"))'` expression | 
| `miner` (3)     | tests targeting miner as primary scenario    | easy filtering tests that validate miner behavior | 
| `signer` (3)    | tests targeting signer as primary scenario   | easy filtering tests that validate signer behavior | 

> **(1)**: `fast` and `parallel` are not really required if we opt to use `not` expression using `nextest`
> So, avoiding them will be good to have less tags (or at least avoiding the ones that could be redundant somehow).
> On the other side, having them allow "easy" filtering (working also with `cargo test`).

> **(2)**: Instead of marking the test with tag `sequential` and handle them with dedicated filter, we could also evaluate the usage of the [serial](https://crates.io/crates/serial_test) crate that enforce test sequential execution at runtime.
> I have used it for some tests in `pinny-rs`([here](https://github.com/BitcoinL2-Labs/pinny-rs/blob/54ba9d533a7b84525a5e65a3eae1a3ae76b9ea49/src/tests/config_test.rs)).
> Obviously tagging them will allow us to extract all tests with that label (using `--list` option on `cargo test` or `nextest`), but we need to explicitly handle runtime execution with proper configuration on the runner.

> **(3)**: Eventually avoidable splitting tests with proper module path (e.g. as signer tests are managed today).

Once this tag proposal becomes consistent, we can update the [contributing](https://github.com/stacks-network/stacks-core/blob/develop/CONTRIBUTING.md#testing) page accordingly.

### Checklist

- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

moodmosaic commented on Cargo.lock:
```diff
@@ -3678,6 +3701,40 @@ dependencies = [
 
+[[package]]
+name = "toml"
+version = "0.8.20"
```
Could we possibly use the existing toml version used elsewhere? (0.5.11) -- It looks like the 0.8.20 one pulls in:
serde_spanned
toml_datetime
toml_edit (which depends on indexmap, winnow)

fdefelici replied:
toml dependency addressed in this commit: 9a2c7bbf60f923f4f653c090407a9cfee6f79a41

moodmosaic reviewed (commented):
This looks good. Thanks for putting it together!

Looking at the test paths:
```
tests::signer::v0::block_proposal_rejection::t::fast::signer::bitcoind::t
```
It feels like the `::t::tag::t` bits are leaking into the visible test paths. Can that tagging stay internal to the library?

fdefelici commented:
> This looks good. Thanks for putting it together!
> 
> Looking at the test paths:
> 
> ```
> tests::signer::v0::block_proposal_rejection::t::fast::signer::bitcoind::t
> ```
> 
> It feels like the `::t::tag::t` bits are leaking into the visible test paths. Can that tagging stay internal to the library?

As explained in the related issue #5975, it is not possible to keep it "internal", because of how `cargo test` and `cargo nextest` works, where test are filtered by paths

moodmosaic commented on Cargo.lock:
```diff
@@ -2202,6 +2202,19 @@ version = "0.1.0"
 
+[[package]]
+name = "pinny"
+version = "0.0.1"
```
Shouldn't this be 0.0.2?

fdefelici replied:
you're right.
fixed here: 9b26d92fda6d01b7abd9a732ad6b71a3419d6509

moodmosaic reviewed (approved):
LGTM!

wileyj commented:
One thing i'd like to see is some additional documentation on how a contributor should implement/use this. 
we don't have any standalone doc for tests/testing - but perhaps just a few lines in the contributing.md (https://github.com/stacks-network/stacks-core/blob/develop/CONTRIBUTING.md#testing)
Or, we could create a new doc under `./docs` that would go into more detail about the testing in stacks-core).

fdefelici commented:
> One thing i'd like to see is some additional documentation on how a contributor should implement/use this. we don't have any standalone doc for tests/testing - but perhaps just a few lines in the contributing.md (https://github.com/stacks-network/stacks-core/blob/develop/CONTRIBUTING.md#testing) Or, we could create a new doc under `./docs` that would go into more detail about the testing in stacks-core).

Updated the contribution page with a light description, eventually to be improved when will have a better tagging integration (tags definition, tests coverage, ...)

Related commit: 5285013e1e89c74ee933d5f658a683fba5797f2e

Does it works for you @wileyj?

fdefelici commented on CONTRIBUTING.md:
```diff
@@ -361,17 +361,24 @@ A non-exhaustive list of examples of consensus-critical changes include:
 
+- Integration tests need to be properly tagged using [pinny-rs](https://github.com/BitcoinL2-Labs/pinny-rs/) crate. Tagging require two fundamental steps:
+  1. Define allowed tags in the package `Cargo.toml` file.
+  2. Use relevant tags for tests, picking from the allowd once
```
fixed here: 81ecd36b8f5f85180d4ba741be06dff8d98134f7

fdefelici commented on testnet/stacks-node/src/tests/signer/v0.rs:
```diff
@@ -29,7 +29,9 @@ use libsigner::v0::messages::{
 use libsigner::{
     BlockProposal, BlockProposalData, SignerSession, StackerDBSession, VERSION_STRING,
 };
+use madhouse::{execute_commands, prop_allof, scenario, Command};
```
I had to resolve a conflict due to the merging of madhouse pr #6007, in this commit: 5d20db1b4cdeff3131f10dccbcf735ca7f462632.

Futhermore I reordered alphabitically the `use` statements to fit `pinny` in between.

obycode commented on CONTRIBUTING.md:
```diff
@@ -361,17 +361,24 @@ A non-exhaustive list of examples of consensus-critical changes include:
 
+- Integration tests need to be properly tagged using [pinny-rs](https://github.com/BitcoinL2-Labs/pinny-rs/) crate. Tagging requires two fundamental steps:
+  1. Define allowed tags in the package `Cargo.toml` file (if needed).
+  2. Apply relevant tags to the tests, picking from the allowed onces.
```
```suggestion
  2. Apply relevant tags to the tests, picking from the allowed set.
```

obycode commented on CONTRIBUTING.md:
```diff
@@ -361,17 +361,24 @@ A non-exhaustive list of examples of consensus-critical changes include:
+  1. Define allowed tags in the package `Cargo.toml` file (if needed).
+  2. Apply relevant tags to the tests, picking from the allowed onces.
+  
+  Then will be possible to run tests with filtering based on the tags using `cargo test` and `cargo nextest` runner.
```
```suggestion
  Then it will be possible to run tests with filtering based on the tags using `cargo test` and `cargo nextest` runner.
```

obycode commented on testnet/stacks-node/src/tests/signer/v0.rs:
```diff
@@ -1487,6 +1488,7 @@ pub fn wait_for_state_machine_update(
     })
 }
 
+#[tag(fast, signer, bitcoind)]
```
How are we defining what is `fast`?

obycode replied:
Oh, I see in your description. Do you plan on putting that table into the documentation?

fdefelici replied:
Yes, that's the idea. I just need to collect our preferences before write it down in the contributing page

hstove reviewed (approved):
Overall, I like the approach! I'd agree about not having a fast/parallel tag.

obycode reviewed (dismissed):
This looks good, assuming the later changes will include documentation on the specific tags we'll use.

fdefelici commented:
- Updated tags configuration to allow `bitcoind`, `flaky`, `slow` tags
- Updated few tests as an example on using the tag
- Updated Contributing page including a table of used tags
- Added a plan (by adding sub-issues to #5974) to puts these tags in action and improving test documentation as well

obycode commented on testnet/stacks-node/src/tests/signer/v0.rs:
```diff
@@ -10122,6 +10122,7 @@ fn multiple_miners_empty_sortition() {
     miners.shutdown();
 }
 
+#[tag(bitcoind, flaky, slow)]
```
I assume you're just putting this here as an example? Let's make sure that once the test runner is updated, we remove `flaky` from this test.

fdefelici replied:
To speak the truth I marked that test as flaky because I pick it from this issue https://github.com/stacks-network/stacks-core/issues/5788 (other than to be an example, my intention was to choose a real one) 

Anyhow, if it is not (or not anymore) , we can took it just as an example and review flaky tests later, as you propose.  
Infact, I already created an adhoc issue to properly tag flaky tests https://github.com/stacks-network/stacks-core/issues/6053  (one of the steps of the proposed plan)
